pywr/parameters/control_curves.py

```diff
@@ -3,79 +3,142 @@
 """
 
 from ._control_curves import BaseControlCurveParameter, ControlCurveInterpolatedParameter
-from .parameters import parameter_registry
+from .parameters import parameter_registry, load_parameter_values, load_parameter
 import numpy as np
 
 
-class ControlCurvePiecewiseParameter(BaseControlCurveParameter):
-    """ A control curve Parameter that returns one of two values depending on whether the current
-    volume is above or below the control curve.
+class ControlCurveParameter(BaseControlCurveParameter):
+    """ A generic multi-levelled control curve Parameter.
+
+     This parameter can be used to return different values when a `Storage` node's current
+      volumes is at different percentage of `max_volume` relative to predefined control curves.
+      Control curves must be defined in the range [0, 1] corresponding to 0% and 100% volume.
+
+     By default this parameter returns an integer sequence from zero if the first control curve
+      is passed, and incrementing by one for each control curve (or "level") the `Storage` node
+      is below.
+
+    Parameters
+    ----------
+    control_curves : `float`, `int` or `Parameter` object, or iterable thereof
+        The position of the control curves. Internally `float` or `int` types are cast to
+        `ConstantParameter`. Multiple values correspond to multiple control curve positions.
+        These should be specified in descending order.
+    values : array_like or `None`, optional
+        The values to return if the `Storage` object is above the correspond control curve.
+        I.e. the first value is returned if the current volume is above the first control curve,
+        and second value if above the second control curve, and so on. The length of `values`
+        must be one more than than the length of `control_curves`. I
+    parameters : iterable `Parameter` objects or `None`, optional
+        If `values` is `None` then `parameters` can specify a `Parameter` object to use at level
+        of the control curves. In the same way as `values` the first `Parameter` is used if
+        `Storage` is above the first control curve, and second `Parameter` if above the
+        second control curve, and so on.
+    storage_node : `Storage` or `None`, optional
```

Comment:
We need to provide the facility to use the combined storage of multiple reservoirs. There are two approaches to this:

1. The control curve needs a `storage_nodes` instead of `storage_node` and the calculations should be done assuming a list of nodes rather than a single node.
2. Add a new "group node" which would behave as a storage node, but underneath would aggregate the percentage calculation.

Both approaches have merit. Option 2 makes the control curve code simpler, and also opens the door for things like group constraints (on non-storage nodes) later (e.g. `group.max_flow = x`). Option 1 can be done within the existing framework. Thoughts?